File: backend/app/routers/export.py
import json
from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import PlainTextResponse, StreamingResponse
from sqlalchemy import select, desc
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models import Project, Source, SocialMention, AnalysisReport
from app.schemas import (
    ProjectResponse,
    SourceResponse,
    SocialMentionResponse,
    AnalysisReportResponse,
    ExportPayload,
)
from app.services.export_service import assemble_export_markdown, assemble_urls_text
from app.services.pdf_service import generate_report_pdf

router = APIRouter(prefix="/api/projects", tags=["export"])


def _project_to_response(p: Project) -> ProjectResponse:
    return ProjectResponse(
        id=p.id,
        chinese_name=p.chinese_name,
        name=p.name or "",
        location=p.location or "",
        project_type=p.project_type or "",
        investors=json.loads(p.investors) if p.investors and p.investors != "[]" else [],
        planning_firms=json.loads(p.planning_firms) if p.planning_firms and p.planning_firms != "[]" else [],
        construction_status=p.construction_status or "unknown",
        operational_data=json.loads(p.operational_data) if p.operational_data and p.operational_data != "{}" else {},
        visitor_count=p.visitor_count or "",
        annual_revenue=p.annual_revenue or "",
        summary=p.summary or "",
        created_at=p.created_at,
        updated_at=p.updated_at,
    )


@router.get("/{project_id}/export", response_model=ExportPayload)
async def export_project(project_id: str, db: AsyncSession = Depends(get_db)):
    """Get full export payload for NotebookLM."""
    stmt = select(Project).where(Project.id == project_id)
    result = await db.execute(stmt)
    project = result.scalar_one_or_none()

    if not project:
        raise HTTPException(status_code=404, detail="项目不存在")

    source_stmt = select(Source).where(Source.project_id == project_id)
    source_result = await db.execute(source_stmt)
    sources = source_result.scalars().all()

    mention_stmt = select(SocialMention).where(SocialMention.project_id == project_id).limit(20)
    mention_result = await db.execute(mention_stmt)
    mentions = mention_result.scalars().all()

    report_stmt = (
        select(AnalysisReport)
        .where(AnalysisReport.project_id == project_id)
        .order_by(desc(AnalysisReport.created_at))
        .limit(1)
    )
    report_result = await db.execute(report_stmt)
    report = report_result.scalar_one_or_none()

    all_urls = [s.url for s in sources]

    return ExportPayload(
        project=_project_to_response(project),
        sources=[SourceResponse(
            id=s.id, project_id=s.project_id, url=s.url, title=s.title or "",
            source_type=s.source_type or "news", platform=s.platform or "tavily",
            snippet=(s.snippet or (s.content or "")[:200]),
            fetch_status=s.fetch_status or "pending", fetched_at=s.fetched_at,
            created_at=s.created_at,
        ) for s in sources],
        social_mentions=[SocialMentionResponse(
            id=m.id, platform=m.platform or "", author=m.author or "",
            content=(m.content or "")[:500], sentiment=m.sentiment or 0.0,
            likes_count=m.likes_count or 0, comments_count=m.comments_count or 0,
            posted_at=m.posted_at, fetched_at=m.fetched_at,
        ) for m in mentions],
        report=AnalysisReportResponse(
            id=report.id, project_id=report.project_id,
            summary=report.summary or "", analysis=report.analysis or "",
            confidence_score=report.confidence_score or 0.0,
            key_findings=json.loads(report.key_findings) if report.key_findings and report.key_findings != "[]" else [],
            recommendations=json.loads(report.recommendations) if report.recommendations and report.recommendations != "[]" else [],
            model_version=report.model_version or "",
            tokens_used=report.tokens_used or 0,
            created_at=report.created_at,
        ) if report else None,
        all_urls=all_urls,
    )


@router.get("/{project_id}/export/markdown", response_class=PlainTextResponse)
async def export_project_markdown(project_id: str, db: AsyncSession = Depends(get_db)):
    """Export project as plain markdown for NotebookLM."""
    stmt = select(Project).where(Project.id == project_id)
    result = await db.execute(stmt)
    project = result.scalar_one_or_none()

    if not project:
        raise HTTPException(status_code=404, detail="项目不存在")

    source_stmt = select(Source).where(Source.project_id == project_id)
    source_result = await db.execute(source_stmt)
    sources = source_result.scalars().all()

    mention_stmt = select(SocialMention).where(SocialMention.project_id == project_id).limit(20)
    mention_result = await db.execute(mention_stmt)
    mentions = mention_result.scalars().all()

    report_stmt = (
        select(AnalysisReport)
        .where(AnalysisReport.project_id == project_id)
        .order_by(desc(AnalysisReport.created_at))
        .limit(1)
    )
    report_result = await db.execute(report_stmt)
    report = report_result.scalar_one_or_none()

    markdown = assemble_export_markdown(project, sources, mentions, report)
    return PlainTextResponse(markdown, media_type="text/plain; charset=utf-8")


@router.get("/{project_id}/export/urls", response_class=PlainTextResponse)
async def export_project_urls(project_id: str, db: AsyncSession = Depends(get_db)):
    """Export all source URLs as text file."""
    stmt = select(Source).where(Source.project_id == project_id)
    result = await db.execute(stmt)
    sources = result.scalars().all()

    if not sources:
        raise HTTPException(status_code=404, detail="没有来源链接")

    return PlainTextResponse(assemble_urls_text(sources), media_type="text/plain; charset=utf-8")


@router.get("/{project_id}/export/pdf")
async def export_project_pdf(project_id: str, db: AsyncSession = Depends(get_db)):
    """Export analysis report as PDF."""
    stmt = select(Project).where(Project.id == project_id)
    result = await db.execute(stmt)
    project = result.scalar_one_or_none()

    if not project:
        raise HTTPException(status_code=404, detail="项目不存在")

    source_stmt = select(Source).where(Source.project_id == project_id)
    source_result = await db.execute(source_stmt)
    sources = source_result.scalars().all()

    report_stmt = (
        select(AnalysisReport)
        .where(AnalysisReport.project_id == project_id)
        .order_by(desc(AnalysisReport.created_at))
        .limit(1)
    )
    report_result = await db.execute(report_stmt)
    report = report_result.scalar_one_or_none()

    if not report:
        raise HTTPException(status_code=404, detail="暂无分析报告，请先生成 AI 分析")

    STATUS_LABELS = {
        "planned": "规划中", "under_construction": "建设中",
        "operating": "运营中", "closed": "已关闭",
    }

    pdf_bytes = generate_report_pdf(
        project_name=project.chinese_name,
        project_info={
            "location": project.location,
            "project_type": project.project_type,
            "construction_status": STATUS_LABELS.get(project.construction_status, project.construction_status),
            "visitor_count": project.visitor_count,
            "annual_revenue": project.annual_revenue,
            "operation_mode": project.operation_mode,
            "investors": json.loads(project.investors) if project.investors and project.investors != "[]" else [],
        },
        analysis_text=report.analysis,
        recommendations=json.loads(report.recommendations) if report.recommendations and report.recommendations != "[]" else [],
        model_version=report.model_version,
        confidence_score=report.confidence_score,
        created_at=report.created_at,
        sources=[{"title": s.title, "url": s.url} for s in sources],
    )

    filename = f"{project.chinese_name}_分析报告.pdf"
    # URL-encode the filename for Content-Disposition
    from urllib.parse import quote
    encoded = quote(filename)

    return StreamingResponse(
        iter([pdf_bytes]),
        media_type="application/pdf",
        headers={
            "Content-Disposition": f"attachment; filename*=UTF-8''{encoded}",
        },
    )

File: frontend/src/lib/export.ts
import { api } from "./api";
import type { ExportPayload } from "@/types";

const STATUS_LABELS: Record<string, string> = {
  planned: "规划中",
  under_construction: "建设中",
  operating: "运营中",
  closed: "已关闭",
  unknown: "未知",
  default: "未知",
};

function sentimentLabel(score: number): string {
  if (score > 0.2) return "😊 正面";
  if (score < -0.2) return "😟 负面";
  return "😐 中性";
}

function assembleMarkdown(payload: ExportPayload): string {
  const lines: string[] = [];

  lines.push(`# ${payload.project.chinese_name} - 文旅大视界项目分析报告`);
  lines.push("");

  lines.push("## 项目信息");
  lines.push(`- **项目名称**: ${payload.project.chinese_name || "待确认"}`);
  if (payload.project.name)
    lines.push(`- **英文名称**: ${payload.project.name}`);
  lines.push(`- **所在地**: ${payload.project.location || "信息收集中"}`);
  lines.push(`- **项目类型**: ${payload.project.project_type || "信息收集中"}`);
  lines.push(
    `- **建设状态**: ${STATUS_LABELS[payload.project.construction_status] || payload.project.construction_status}`
  );
  if (payload.project.investors?.length)
    lines.push(`- **投资方**: ${payload.project.investors.join("、")}`);
  if (payload.project.planning_firms?.length)
    lines.push(
      `- **规划设计**: ${payload.project.planning_firms.join("、")}`
    );
  if (payload.project.visitor_count)
    lines.push(`- **年接待游客**: ${payload.project.visitor_count}`);
  if (payload.project.annual_revenue)
    lines.push(`- **年收入**: ${payload.project.annual_revenue}`);
  lines.push("");

  lines.push("## 来源链接");
  lines.push("");
  payload.sources.forEach((s, i) => {
    const title = s.title || "来源";
    lines.push(`${i + 1}. ${title}`);
    lines.push(`   ${s.url}`);
    lines.push("");
  });

  if (payload.social_mentions?.length) {
    lines.push("## 社交媒体评价");
    lines.push("以下为社交媒体平台收集的用户评价：");
    lines.push("");
    for (const m of payload.social_mentions) {
      const preview = (m.content || "").slice(0, 200);
      lines.push(
        `- **[${m.platform}]** ${m.author || "匿名用户"}: ${preview}`
      );
      lines.push(
        `  - 情感倾向: ${sentimentLabel(m.sentiment)} | ❤️ ${m.likes_count} | 💬 ${m.comments_count}`
      );
    }
    lines.push("");
  }

  if (payload.report) {
    lines.push("## AI 分析报告");
    lines.push("");
    lines.push(payload.report.analysis);
    lines.push("");
    lines.push("---");
    lines.push(
      `*报告生成时间: ${new Date(payload.report.created_at).toLocaleString("zh-CN")}*`
    );
    lines.push(
      `*置信度: ${(payload.report.confidence_score * 100).toFixed(0)}%*`
    );
    lines.push(`*分析模型: ${payload.report.model_version}*`);
  }

  return lines.join("\n");
}

export async function copyToClipboard(projectId: string): Promise<string> {
  const payload: ExportPayload = await api.getExportPayload(projectId);
  const markdown = assembleMarkdown(payload);
  await navigator.clipboard.writeText(markdown);
  return markdown;
}

export async function copyReportOnly(projectId: string): Promise<string> {
  const payload: ExportPayload = await api.getExportPayload(projectId);
  if (!payload.report) {
    await navigator.clipboard.writeText("暂无AI分析报告");
    return "暂无AI分析报告";
  }
  const reportText = `# ${payload.project.chinese_name} - AI 分析报告\n\n${payload.report.analysis}\n\n---\n*报告生成时间: ${new Date(payload.report.created_at).toLocaleString("zh-CN")}*\n*分析模型: ${payload.report.model_version}*`;
  await navigator.clipboard.writeText(reportText);
  return reportText;
}

export async function copySourcesOnly(projectId: string): Promise<string> {
  const payload: ExportPayload = await api.getExportPayload(projectId);
  const lines: string[] = [];
  lines.push(`# ${payload.project.chinese_name} - 来源链接`);
  lines.push("");
  payload.sources.forEach((s, i) => {
    const title = s.title || "来源";
    lines.push(`${i + 1}. ${title}`);
    lines.push(`   ${s.url}`);
    lines.push("");
  });
  const sourceText = lines.join("\n");
  await navigator.clipboard.writeText(sourceText);
  return sourceText;
}

export function openNotebookLM(): void {
  window.open("https://notebooklm.google.com", "_blank");
}

export async function copyAndOpenNotebookLM(projectId: string): Promise<void> {
  const markdown = await copyToClipboard(projectId);
  openNotebookLM();
}
